src/QsCompiler/Compiler/RewriteSteps/CapabilityInference.cs

```diff
@@ -0,0 +1,42 @@
+﻿#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Quantum.QsCompiler.SyntaxTree;
+using static Microsoft.Quantum.QsCompiler.SyntaxProcessing.CapabilityInference;
+
+namespace Microsoft.Quantum.QsCompiler.BuiltInRewriteSteps
+{
+    /// <summary>
+    /// Infers the minimum runtime capabilities required by each callable.
+    /// </summary>
+    internal class CapabilityInference : IRewriteStep
+    {
+        public string Name { get; } = "Capability Inference";
+
+        public int Priority { get; } = RewriteStepPriorities.CapabilityInference;
+
+        public IDictionary<string, string?> AssemblyConstants { get; } =
+            new Dictionary<string, string?>();
+
+        public IEnumerable<IRewriteStep.Diagnostic> GeneratedDiagnostics { get; } =
+            Enumerable.Empty<IRewriteStep.Diagnostic>();
+
+        public bool ImplementsPreconditionVerification { get; } = false;
+
+        public bool ImplementsTransformation { get; } = true;
+
+        public bool ImplementsPostconditionVerification { get; } = false;
+
+        public bool PreconditionVerification(QsCompilation compilation) => throw new NotSupportedException();
```

Comment:
It's not valid for a Q# library to have an execution target, right?

Reply:
Yes, that is correct.

Reply:
Shouldn't rewrite steps be able to assume certain correctness properties about the compilation, such as that a library compilation can't have an execution target, rather than needing to check for it explicitly?

Reply:
The rewrite steps should certainly assume that the QsCompilation object is valid, and I think also assuming that the assembly constants passed by the compiler are accurate seems fair. But that is different from relying on knowledge about the relation that assembly constants should have (you did ask the question about whether or not a Q# library can be targeted for example). 
Aside from that, you may be right in that the capability inference should run even if it is a targeted compilation. So I am good with keeping it as is.

Reply:
I can see it being useful as a sanity test. I added this as a precondition, and also that the `RequiresCapability` attribute must be defined in the compilation.